src/test/kotlin/org/rust/cargo/runconfig/TestRunnerTest.kt

```diff
@@ -0,0 +1,370 @@
+        assert(test2.duration!! < 100)
+
+        val mod = findTestByName("tests", root)!!
+        assert(mod.duration!! == test1.duration!! + test2.duration!!)
+    }
+
+    private fun createTestRunConfigurationFromContext(): CargoCommandConfiguration {
+        val context = DataManager.getInstance().getDataContext(myFixture.editor.component)
+        return CargoTestRunConfigurationProducer()
+            .createConfigurationFromContext(ConfigurationContext.getFromContext(context))
+            ?.configuration as? CargoCommandConfiguration
+            ?: error("Can't create run configuration")
+    }
+
+    private fun runConfigurationAndGetTestRoot(configuration: RunConfiguration): SMTestProxy.SMRootTestProxy {
+        val executor = DefaultRunExecutor.getRunExecutorInstance()
+        val state = ExecutionEnvironmentBuilder
+            .create(executor, configuration)
+            .build()
+            .state!!
+        val result = state.execute(executor, RsRunner())!!
+        val executionConsole = result.executionConsole as SMTRunnerConsoleView
+        val testsRootNode = executionConsole.resultsViewer.testsRootNode
+
+        with(result.processHandler) {
+            startNotify()
+            waitFor()
+        }
+        LaterInvocator.dispatchPendingFlushes()
+        Disposer.dispose(executionConsole)
+
+        return testsRootNode
+    }
+
+    private fun runTestsAndCheckTestTree(expectedFormattedTestTree: String) {
+        val configuration = createTestRunConfigurationFromContext()
+        val root = runConfigurationAndGetTestRoot(configuration)
+        assertEquals(expectedFormattedTestTree.trimIndent(), getFormattedTestTree(root))
+    }
+
+    companion object {
+        private fun getFormattedTestTree(testTreeRoot: SMTestProxy.SMRootTestProxy): String =
+            buildString {
+                if (testTreeRoot.wasTerminated()) {
+                    append("Test terminated")
+                    return@buildString
+                }
+                formatLevel(testTreeRoot)
+            }
+
+        private fun StringBuilder.formatLevel(test: SMTestProxy, level: Int = 0) {
+            append(".".repeat(level))
+            append(test.name)
+            when {
+                test.wasTerminated() -> append("[T]")
+                test.isPassed -> append("(+)")
+                test.isIgnored -> append("(~)")
+                else -> append("(-)")
+            }
+
+            for (child in test.children) {
+                append('\n')
+                formatLevel(child, level + 1)
+            }
+        }
+
+        private fun findTestByName(testFullName: String, root: SMTestProxy.SMRootTestProxy): SMTestProxy? {
+            for (child in root.children) {
+                val result = findTestByName(testFullName, child)
+                if (result != null) return result
+            }
+            return null
+        }
+
+        private fun findTestByName(
```

Comment:
Personally, I usually prefer to encapsulate recursive part of similar functions in inner function
I mean something like
```kotlin
private fun findTestByName(testFullName: String, root: SMTestProxy.SMRootTestProxy): SMTestProxy? {
    val fullNameBuffer = mutableListOf<String>()

    fun find(test: SMTestProxy): SMTestProxy? {
        if (test !is TestProxyRoot) {
            fullNameBuffer.add(test.name)
        }
        if (testFullName == fullNameBuffer.joinToString("::")) return test
        for (child in test.children) {
            val result = find(child)
            if (result != null) return result
        }
        if (test !is TestProxyRoot) {
            fullNameBuffer.removeLast()
        }
        return null
    }

    return find(root)
}
```

Just a comment, I don't insist 😄